x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java

```diff
@@ -0,0 +1,199 @@
+import org.elasticsearch.xpack.core.security.user.User;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static org.elasticsearch.xpack.core.security.authc.Authentication.VERSION_API_KEY_ROLES_AS_BYTES;
+import static org.elasticsearch.xpack.core.security.authc.AuthenticationField.API_KEY_LIMITED_ROLE_DESCRIPTORS_KEY;
+import static org.elasticsearch.xpack.core.security.authc.AuthenticationField.API_KEY_ROLE_DESCRIPTORS_KEY;
+
+/**
+ * A subject is a more generic concept similar to user and associated to the current authentication.
+ * It is more generic than user because it can also represent API keys and service accounts.
+ * It also contains authentication level information, e.g. realm and metadata so that it can answer
+ * queries in a better encapsulated way.
+ */
+public class Subject {
+
+    public enum Type {
+        USER,
+        API_KEY,
+        SERVICE_ACCOUNT,
+    }
+
+    private final Version version;
+    private final User user;
+    private final Authentication.RealmRef realm;
+    private final Type type;
+    private final Map<String, Object> metadata;
+
+    public Subject(User user, Authentication.RealmRef realm) {
+        this(user, realm, Version.CURRENT, Map.of());
+    }
+
+    public Subject(User user, Authentication.RealmRef realm, Version version, Map<String, Object> metadata) {
+        this.version = version;
+        this.user = user;
+        this.realm = realm;
+        // Realm can be null for run-as user if it does not exist. Pretend it is a user and it will be rejected later in authorization
+        // This is to be consistent with existing behaviour.
+        if (realm == null) {
+            this.type = Type.USER;
+        } else if (AuthenticationField.API_KEY_REALM_TYPE.equals(realm.getType())) {
+            assert AuthenticationField.API_KEY_REALM_NAME.equals(realm.getName()) : "api key realm name mismatch";
+            this.type = Type.API_KEY;
+        } else if (ServiceAccountSettings.REALM_TYPE.equals(realm.getType())) {
+            assert ServiceAccountSettings.REALM_NAME.equals(realm.getName()) : "service account realm name mismatch";
+            this.type = Type.SERVICE_ACCOUNT;
+        } else {
+            this.type = Type.USER;
+        }
+        this.metadata = metadata;
+    }
+
+    public User getUser() {
+        return user;
+    }
+
+    public Authentication.RealmRef getRealm() {
+        return realm;
+    }
+
+    public Type getType() {
+        return type;
+    }
+
+    public Map<String, Object> getMetadata() {
+        return metadata;
+    }
+
+    /**
+     * Return a List of RoleReferences that represents role definitions associated to the subject.
+     * The final role of this subject should be the intersection of all role references in the list.
+     */
+    public List<RoleReference> getRoleReferences(@Nullable AnonymousUser anonymousUser) {
```

Comment:
In the eariler version, I wrapped the List inside a dedicate class. But decided later to just return the list directly. Because (1) I didn't come up with other useful field or method for that class; (2) the usage of the list is quite simple, it must be intersected and order not does not even matter.

Reply:
I prefer a separate class, but we can handle it as a separate discussion.
My main reason is that I don't think a `List` means anything without a lot of context, but `RoleIntersection` does.

However, if we're going to do a separate PR to handle `LimitedRole` etc, we can bundle up a series of changes there.

Reply:
Makes sense to me. I raised https://github.com/elastic/elasticsearch/issues/81192 and let's sort it out there.